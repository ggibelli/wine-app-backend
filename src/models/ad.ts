/* eslint-disable @typescript-eslint/no-unsafe-member-access */
import mongoose, { Schema, Document } from 'mongoose';
// import { Address } from '../types';
import {
  // Province,
  // Regioni,
  METODOPRODUTTIVO,
} from '../utils/enumMongooseHelper';
import { IUserDoc } from './user';
import { IWineDoc } from './wine';
import { TypeAd, TypeProduct, Menzione } from '../types';
import { IVineyardDoc } from './vineyard';
import { MetodoProduttivo } from '../types';
import { Negotiation } from '../generated/graphql';

export interface IAd {
  postedBy: IUserDoc['_id'];
  wineName?: string;
  wine?: IWineDoc['_id'];
  sottoZona?: string;
  menzione?: Menzione;
  metodoProduttivo?: MetodoProduttivo;
  vineyardName?: string;
  vineyard?: IVineyardDoc['_id'];
  harvest: number;
  abv: number;
  priceFrom: number; //se vendita priceFrom e priceTo settati a stesso numero
  priceTo: number;
  litersFrom?: number; //se vendite litersFrom e litersTo settati a stesso numero
  litersTo?: number;
  kgFrom?: number;
  kgTo?: number;
  content?: string | null | undefined;
  // address: Address;
  negotiations?: mongoose.Types.Array<Negotiation>; // trattative dell'annuncio, solo attive, solo graphql??
  viewedBy?: mongoose.Types.Array<IUserDoc['_id']>;
  savedBy?: mongoose.Types.Array<IUserDoc['_id']>;
  typeAd: TypeAd;
  typeProduct: TypeProduct;
  needsFollowUp: boolean;
  isActive: boolean;
  datePosted: Date;
}

export interface AdGraphQl extends IAd {
  _id: mongoose.Types.ObjectId | string;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const adSchemaFields: Record<keyof IAd, any> = {
  postedBy: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  typeAd: {
    type: String,
    enum: [TypeAd.SELL, TypeAd.BUY],
    required: true,
  },
  typeProduct: {
    type: String,
    enum: [TypeProduct.ADWINE, TypeProduct.ADGRAPE],
    required: true,
  },
  wineName: {
    type: String,
    minlength: 3,
    required() {
      return this.typeProduct === TypeProduct.ADWINE;
    },
    index: true,
  },
  wine: {
    type: Schema.Types.ObjectId,
    ref: 'Wine',
    // required() {
    //   return this.typeProduct === TypeProduct.ADWINE;
    // },
  },
  sottoZona: {
    type: String,
  },
  menzione: {
    type: String,
    enum: ['CLASSICO', 'RISERVA', 'SUPERIORE', 'VIGNA'],
  },
  metodoProduttivo: {
    type: String,
    enum: Object.values(METODOPRODUTTIVO),
    required: true,
    default: 'CONVENZIONALE',
  },
  vineyardName: {
    type: String,
    minlength: 3,
    required() {
      return this.typeProduct === TypeProduct.ADGRAPE;
    },
    index: true,
  },
  vineyard: {
    type: Schema.Types.ObjectId,
    ref: 'Vineyard',
    // required() {
    //   return this.typeProduct === TypeProduct.ADGRAPE;
    // },
  },
  abv: Number,
  harvest: {
    type: Number,
    required: true,
  },
  priceFrom: {
    type: Number,
    required: true,
  },
  priceTo: {
    type: Number,
    required: true,
  },
  litersFrom: {
    type: Number,
    required() {
      return this.typeProduct === TypeProduct.ADWINE;
    },
  },
  litersTo: {
    type: Number,
    required() {
      return this.typeProduct === TypeProduct.ADWINE;
    },
  },
  kgFrom: {
    type: Number,
    required() {
      return this.typeProduct === TypeProduct.ADGRAPE;
    },
  },
  kgTo: {
    type: Number,
    required() {
      return this.typeProduct === TypeProduct.ADGRAPE;
    },
  },
  // address: {
  //   comune: {
  //     type: String,
  //     required: true,
  //   },
  //   provincia: {
  //     type: String,
  //     enum: Object.values(Province),
  //     required: true,
  //   },
  //   regione: {
  //     type: String,
  //     enum: Object.values(Regioni),
  //     required: true,
  //   },
  // },
  negotiations: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Negotiation',
    },
  ],
  viewedBy: [
    {
      type: Schema.Types.ObjectId,
      ref: 'User',
    },
  ],
  savedBy: [
    {
      type: Schema.Types.ObjectId,
      ref: 'User',
    },
  ],
  content: {
    type: String,
  },
  isActive: {
    type: Boolean,
    default: true,
  },
  needsFollowUp: {
    type: Boolean,
    default: true,
  },
  datePosted: {
    type: Date,
    // `Date.now()` returns the current unix timestamp as a number
    default: Date.now,
  },
};

export interface IAdDoc extends IAd, Document {}

const adSchema = new Schema(adSchemaFields);

adSchema.index({ typeProduct: 1, typeAd: 1, wineName: 1, vineyardName: 1 });

export const Ad = mongoose.model<IAdDoc>('Ad', adSchema);
