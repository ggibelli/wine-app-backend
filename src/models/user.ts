/* eslint-disable @typescript-eslint/no-unsafe-return */
import mongoose, { Schema, Document, HookNextFunction } from 'mongoose';
import mongooseUniqueValidator from 'mongoose-unique-validator';
import { Address } from '../types';
import { Province, Regioni } from '../utils/enumMongooseHelper';
import { IAdDoc } from './ad';
import { IWineDoc } from './wine';
import { IReviewDoc } from './review';
import { IVineyardDoc } from './vineyard';
import { MetodoProduttivo } from '../types';
import { METODOPRODUTTIVO } from '../utils/enumMongooseHelper';
import bcrypt from 'bcrypt';
import { IMessageDoc } from './message';
import { Negotiation } from '../generated/graphql';

const HASH_ROUNDS = 10;

interface ProducedWines {
  wine: IWineDoc['_id'] | IWineDoc;
  bottlesProduced: number;
  metodoProduttivo: MetodoProduttivo;
}

interface OwnedVineyards {
  vineyard: IVineyardDoc['_id'] | IVineyardDoc;
  tonsProduced: number;
  metodoProduttivo: MetodoProduttivo;
}

export interface Coordinates {
  latitude: number;
  longitude: number;
}

export interface IUser {
  email: string;
  firstName: string;
  lastName: string;
  password: string;
  pIva: string;
  phoneNumber: string;
  address: Address;
  coordinates?: Coordinates;
  isVerified: boolean;
  isPremium?: boolean;
  isAdmin: boolean;
  hideContact: boolean;
  ads?: mongoose.Types.Array<IAdDoc['_id'] | IAdDoc>; // annunci postati dall'utente
  savedAds?: mongoose.Types.Array<IAdDoc['_id'] | IAdDoc>; // annunci postati dall'utente
  messages?: mongoose.Types.Array<IMessageDoc['_id'] | IMessageDoc>;
  negotiations?: mongoose.Types.Array<Negotiation>; // trattative dell'utente, attive e non, concluse e non
  reviews?: mongoose.Types.Array<IReviewDoc['_id'] | IReviewDoc>; // recensioni fatte e ricevute dall'utente
  adsRemaining?: number;
  dateCreated: Date;
  producedWines?: mongoose.Types.Array<ProducedWines>;
  ownedVineyards?: mongoose.Types.Array<OwnedVineyards>;
}

export interface IUserDoc extends IUser, Document {
  validatePassword(password: string): Promise<boolean>;
}

export interface UserGraphQl extends IUser {
  validatePassword(password: string): Promise<boolean>;
  _id: mongoose.Types.ObjectId;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const userSchemaFields: Record<keyof IUser, any> = {
  email: {
    type: String,
    unique: true,
    minlength: 3,
    required: true,
  },
  phoneNumber: {
    type: String,
    unique: true,
    minlength: 10,
    required: true,
  },
  firstName: String,
  lastName: String,
  password: String,
  pIva: {
    type: String,
    unique: true,
    minlength: 11,
    required: true,
  },
  address: {
    via: {
      type: String,
      required: true,
      minlength: 5,
    },
    CAP: {
      type: String,
      required: true,
      minlength: 5,
    },
    comune: { type: String, required: true, minlength: 5 },
    provincia: {
      type: String,
      enum: Object.values(Province),
      required: true,
    },
    regione: {
      type: String,
      enum: Object.values(Regioni),
      required: true,
    },
  },
  coordinates: {
    latitude: Number,
    longitude: Number,
  },
  isVerified: {
    type: Boolean,
    required: true,
    default: false,
  },
  isPremium: {
    type: Boolean,
    default: false,
  },
  isAdmin: {
    type: Boolean,
    default: false,
  },
  hideContact: {
    type: Boolean,
    default: true,
  },
  ads: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Ad',
    },
  ],
  savedAds: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Ad',
    },
  ],
  messages: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Message',
    },
  ],
  negotiations: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Negotiation',
    },
  ],
  reviews: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Review',
    },
  ],
  adsRemaining: Number,
  dateCreated: {
    type: Date,
    // `Date.now()` returns the current unix timestamp as a number
    default: Date.now,
  },
  producedWines: [
    {
      wine: {
        type: Schema.Types.ObjectId,
        ref: 'Wine',
      },
      bottlesProduced: Number,
      metodoProduttivo: {
        type: String,
        enum: Object.values(METODOPRODUTTIVO),
      },
    },
  ],
  ownedVineyards: [
    {
      vineyard: {
        type: Schema.Types.ObjectId,
        ref: 'Vineyard',
      },
      tonsProduced: Number,
      metodoProduttivo: {
        type: String,
        enum: Object.values(METODOPRODUTTIVO),
      },
    },
  ],
};

const userSchema = new Schema(userSchemaFields);

userSchema.pre('save', async function (next: HookNextFunction) {
  // here we need to retype 'this' because by default it is
  // of type Document from which the 'IUser' interface is inheriting
  // but the Document does not know about our password property
  const thisObj = this as IUserDoc;

  if (!this.isModified('password')) {
    return next();
  }

  try {
    const salt = await bcrypt.genSalt(HASH_ROUNDS);
    thisObj.password = await bcrypt.hash(thisObj.password, salt);
    return next();
  } catch (e) {
    return next(e);
  }
});

userSchema.methods.validatePassword = async function (pass: string) {
  return bcrypt.compare(pass, this.password);
};

userSchema.plugin(mongooseUniqueValidator);

export const User = mongoose.model<IUserDoc>('User', userSchema);
