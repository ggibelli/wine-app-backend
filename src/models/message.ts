import mongoose, { Schema, Document } from 'mongoose';
import { IUserDoc } from './user';
import { INegotiationDoc } from './negotiation';

export interface IMessage {
  negotiation: INegotiationDoc['_id'];
  content: string;
  sentBy: IUserDoc['_id'];
  sentTo: IUserDoc['_id'];
  dateSent: Date;
  isViewed: boolean;
}

export interface MessageGraphQl extends IMessage {
  _id: mongoose.Types.ObjectId;
}

export interface IMessageDoc extends IMessage, Document {}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const messageSchemaFields: Record<keyof IMessage, any> = {
  negotiation: {
    type: Schema.Types.ObjectId,
    ref: 'Negotiation',
    required: true,
  },
  content: {
    type: String,
    minlength: 1,
    required: true,
  },
  sentBy: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  sentTo: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  dateSent: {
    type: Date,
    // `Date.now()` returns the current unix timestamp as a number
    default: Date.now,
  },
  isViewed: {
    type: Boolean,
    default: false,
  },
};

const messageSchema = new Schema(messageSchemaFields);

messageSchema.index({ negotiation: 1, sentTo: 1, sentFrom: 1 });

export const Message = mongoose.model<IMessageDoc>('Message', messageSchema);
