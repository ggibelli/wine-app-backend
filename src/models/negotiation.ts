import mongoose, { Schema, Document } from 'mongoose';
import { IUserDoc } from './user';
import { IAdDoc } from './ad';
import { IMessageDoc } from './message';
import { TypeAd } from '../types';
import mongooseUniqueValidator from 'mongoose-unique-validator';

export interface INegotiation {
  createdBy: IUserDoc['_id'];
  ad: IAdDoc['_id'];
  forUserAd: IUserDoc['_id'];
  messages?: mongoose.Types.Array<IMessageDoc['_id']>;
  isConcluded: boolean;
  dateCreated: Date;
  type: TypeAd;
}

export interface NegotiationGraphQl extends INegotiation {
  _id: mongoose.Types.ObjectId;
}

export interface INegotiationDoc extends INegotiation, Document {}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const negotiationSchemaFields: Record<keyof INegotiation, any> = {
  createdBy: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  type: {
    type: String,
    enum: ['SELL', 'BUY'],
    required: true,
  },
  ad: {
    type: Schema.Types.ObjectId,
    ref: 'Ad',
    required: true,
  },
  forUserAd: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  messages: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Message',
    },
  ],
  isConcluded: {
    type: Boolean,
    default: false,
  },
  dateCreated: {
    type: Date,
    // `Date.now()` returns the current unix timestamp as a number
    default: Date.now,
  },
};

const negotiationSchema = new Schema(negotiationSchemaFields);

negotiationSchema.index({ createdBy: 1, ad: 1 }, { unique: true });
negotiationSchema.plugin(mongooseUniqueValidator);
export const Negotiation = mongoose.model<INegotiationDoc>(
  'Negotiation',
  negotiationSchema
);
