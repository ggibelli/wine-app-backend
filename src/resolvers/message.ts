/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-call */
import {
  Resolvers,
  MessageInput,
  //MessageInputUpdate,
} from '../generated/graphql';
import Users from '../data-sources/users';
import Messages from '../data-sources/messages';
import Negotiations from '../data-sources/negotiations';
import { MessageGraphQl } from '../models/message';
import { UserGraphQl } from '../models/user';
import { PubSub, withFilter } from 'apollo-server-express';

const pubsub = new PubSub();

const MESSAGE_SENT = 'MESSAGE_SENT';

interface StringIndexSignatureInterface {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [index: string]: any;
}

type StringIndexed<T> = T & StringIndexSignatureInterface;

interface MongoDataSource {
  users: Users;
  messages: Messages;
  negotiations: Negotiations;
}

export const resolver: StringIndexed<Resolvers> = {
  Query: {
    async messagesToUser(
      _,
      { sentTo },
      { dataSources }: { dataSources: MongoDataSource }
    ) {
      return dataSources.messages.getMessagesTo(sentTo);
    },
    async messagesForNegotiation(
      _,
      { negotiation },
      { dataSources }: { dataSources: MongoDataSource }
    ) {
      return dataSources.messages.getMessagesForNegotiation(negotiation);
    },
    async messages(_, __, { dataSources }: { dataSources: MongoDataSource }) {
      return dataSources.messages.getMessages();
    },
    async message(
      _,
      { id },
      { dataSources }: { dataSources: MongoDataSource }
    ) {
      return dataSources.messages.getMessage(id);
    },
  },
  Mutation: {
    async createMessage(
      _,
      { message }: { message: MessageInput },
      { dataSources, user }: { dataSources: MongoDataSource; user: UserGraphQl }
    ) {
      const negotiation = await dataSources.negotiations.getNegotiation(
        message.negotiation
      );
      if (
        negotiation?.createdBy.toString() !== user._id.toString() &&
        negotiation?.forUserAd.toString() !== user._id.toString()
      ) {
        return {
          response: null,
          errors: [
            {
              name: 'UserInputError',
              text: 'You can only send messages if you own the negotiation',
            },
          ],
        };
      }
      const messageResponse = await dataSources.messages.createMessage(message);
      if (!messageResponse.errors.length) {
        await pubsub.publish(MESSAGE_SENT, {
          messageSent: messageResponse.response,
        });
      }
      return messageResponse;
    },

    // updateMessage(
    //   _,
    //   { message }: { message: MessageInputUpdate },
    //   { dataSources }: { dataSources: MongoDataSource }
    // ) {
    //   return dataSources.messages.updateMessage(message);
    // },
    // deleteMessage(
    //   _: void,
    //   { id }: { id: string },
    //   { dataSources }: { dataSources: MongoDataSource }
    // ) {
    //   return dataSources.messages.deleteMessage(id);
    // },
  },

  Subscription: {
    messageSent: {
      subscribe: withFilter(
        () => pubsub.asyncIterator([MESSAGE_SENT]),
        (
          payload: { messageSent: MessageGraphQl },
          _,
          { user }: { user: UserGraphQl }
        ) => {
          return Boolean(
            payload.messageSent.sentTo.toString() === user._id.toString()
          );
        }
      ),
    },
  },

  Message: {
    async sentBy(
      message: MessageGraphQl,
      _,
      { dataSources }: { dataSources: MongoDataSource }
    ) {
      return dataSources.users.getUser(message.sentBy);
    },
    async sentTo(
      message: MessageGraphQl,
      _,
      { dataSources }: { dataSources: MongoDataSource }
    ) {
      return dataSources.users.getUser(message.sentTo);
    },
    async negotiation(
      message: MessageGraphQl,
      _,
      { dataSources }: { dataSources: MongoDataSource }
    ) {
      return dataSources.negotiations.getNegotiation(message.negotiation);
    },
  },
};
