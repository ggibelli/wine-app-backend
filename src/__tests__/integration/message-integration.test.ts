/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import { gql } from 'apollo-server-express';
import { LeanDocument, Types } from 'mongoose';
import {
  testClient,
  connectToDb,
  dropTestDb,
  closeDbConnection,
} from '../../tests/integrationSetup';
// import { ObjectId } from 'mongodb';
import { User } from '../../models/user';
import { Ad } from '../../models/ad';
import { users, ads } from '../../tests/mocksTests';
import { MessageInput } from '../../generated/graphql';

import { Negotiation } from '../../models/negotiation';
import { Message } from '../../models/message';
import {
  NegotiationDocument,
  MessageDocument,
} from '../../interfaces/mongoose.gen';

const { query, mutate, setOptions } = testClient;

const MESSAGES_USER = gql`
  query GetMessagesUser($user: ID!) {
    messagesToUser(sentTo: $user) {
      sentBy {
        firstName
      }
      sentTo {
        firstName
      }
      content
    }
  }
`;

const MESSAGES_NEGOTIATION = gql`
  query GetMessagesNegotiation($negotiation: ID!) {
    messagesForNegotiation(negotiation: $negotiation) {
      messages {
        sentBy {
          firstName
        }
        sentTo {
          firstName
        }
        content
        negotiation {
          createdBy {
            firstName
          }
          forUserAd {
            firstName
          }
        }
      }
    }
  }
`;

const MESSAGES = gql`
  {
    messages {
      sentBy {
        firstName
      }
      sentTo {
        firstName
      }
      content
    }
  }
`;

const MESSAGE = gql`
  query GetMessage($id: ID!) {
    message(id: $id) {
      sentBy {
        firstName
      }
      sentTo {
        firstName
      }
      content
    }
  }
`;

const CREATE_MESSAGE = gql`
  mutation CreateMessage($message: MessageInput!) {
    createMessage(message: $message) {
      response {
        sentBy {
          firstName
        }
        sentTo {
          firstName
        }
        content
        negotiation {
          forUserAd {
            firstName
          }
          createdBy {
            firstName
          }
        }
      }
      errors {
        name
        text
      }
    }
  }
`;

const LOGIN_VALID = gql`
  mutation {
    login(email: "gio@prova.it", password: "giovanni") {
      response {
        token
      }
      errors {
        name
        text
      }
    }
  }
`;

const LOGIN_VALID_OTHER = gql`
  mutation {
    login(email: "gio2@prova.it", password: "giovanni") {
      response {
        token
      }
      errors {
        name
        text
      }
    }
  }
`;

beforeAll(async () => {
  await connectToDb();
  await dropTestDb();
  const usersMock = users();
  const adsMock = ads();
  const user = new User({ ...usersMock[0], _id: new Types.ObjectId() });
  const otherUser = new User({ ...usersMock[1], _id: new Types.ObjectId() });
  const thirdUser = new User({ ...usersMock[2], _id: new Types.ObjectId() });
  const ad = new Ad({
    ...adsMock[0],
    postedBy: otherUser,
    _id: new Types.ObjectId(),
  });
  const negotiation = new Negotiation({
    createdBy: user,
    ad,
    forUserAd: ad.postedBy,
    type: ad.typeAd,
    _id: new Types.ObjectId(),
  });
  const otherNegotiation = new Negotiation({
    createdBy: thirdUser,
    ad,
    forUserAd: ad.postedBy,
    type: ad.typeAd,
    _id: new Types.ObjectId(),
  });
  const message = new Message({
    negotiation,
    sentBy: negotiation.createdBy,
    sentTo: negotiation.forUserAd,
    content: 'ciao',
    _id: new Types.ObjectId(),
  });
  const otherMessage = new Message({
    negotiation: otherNegotiation,
    sentBy: otherNegotiation.createdBy,
    sentTo: otherNegotiation.forUserAd,
    content: 'ciao di nuovo',
    _id: new Types.ObjectId(),
  });
  await message.save();
  await otherMessage.save();
  await user.save();
  await otherUser.save();
  await thirdUser.save();
  await ad.save();
  await negotiation.save();
  await otherNegotiation.save();
});

describe('Integration test messages', () => {
  it('query message fails if not logged in', async () => {
    const res = await query(MESSAGES);
    expect(res).toMatchSnapshot();
  });

  it('query messagesToUser fails if not logged in', async () => {
    const negotiation: LeanDocument<NegotiationDocument>[] =
      await Negotiation.find({}).lean().exec();
    const res = await query(MESSAGES_USER, {
      variables: { user: negotiation[0].forUserAd.toString() },
    });
    expect(res).toMatchSnapshot();
  });

  it('query messageForNegotiation fails if not logged in', async () => {
    const negotiation: LeanDocument<NegotiationDocument>[] =
      await Negotiation.find({}).lean().exec();
    const res = await query(MESSAGES_NEGOTIATION, {
      variables: { negotiation: negotiation[0]._id.toString() },
    });
    expect(res).toMatchSnapshot();
  });

  it('query single message fails if not logged', async () => {
    const message: LeanDocument<MessageDocument>[] = await Message.find({})
      .lean()
      .exec();
    const res = await query(MESSAGE, {
      variables: { id: message[0]._id.toString() },
    });
    expect(res).toMatchSnapshot();
  });

  it('create message mutation fails if not logged in', async () => {
    const negotiation: LeanDocument<NegotiationDocument>[] =
      await Negotiation.find({}).lean().exec();
    const message: MessageInput = {
      negotiation: negotiation[0]._id.toString(),
      sentTo: negotiation[0].forUserAd.toString(),
      content: 'prova prova',
    };
    const res = await mutate(CREATE_MESSAGE, {
      variables: { message },
    });
    expect(res).toMatchSnapshot();
  });

  it('create message mutation successfull', async () => {
    const data: any = await mutate(LOGIN_VALID);
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    const { token } = data.data.login.response;
    setOptions({
      request: {
        headers: {
          authorization: token,
        },
      },
    });
    const negotiation: LeanDocument<NegotiationDocument>[] =
      await Negotiation.find({}).lean().exec();
    const message: MessageInput = {
      negotiation: negotiation[0]._id.toString(),
      sentTo: negotiation[0].forUserAd.toString(),
      content: 'prova prova',
    };
    const res = await mutate(CREATE_MESSAGE, {
      variables: { message },
    });
    expect(res).toMatchSnapshot();
  });

  it('create message mutation fails if user does not own the negotiation', async () => {
    const negotiation: LeanDocument<NegotiationDocument>[] =
      await Negotiation.find({}).lean().exec();
    const message: MessageInput = {
      negotiation: negotiation[1]._id.toString(),
      sentTo: negotiation[1].forUserAd.toString(),
      content: 'prova prova',
    };
    const res = await mutate(CREATE_MESSAGE, {
      variables: { message },
    });
    expect(res).toMatchSnapshot();
  });

  it('create message mutation fails if recipient same as sender', async () => {
    const user = await User.findOne({ firstName: 'Giovanni' }).lean().exec();
    if (!user) throw new Error();
    const negotiation: LeanDocument<NegotiationDocument>[] =
      await Negotiation.find({}).lean().exec();
    const message: MessageInput = {
      negotiation: negotiation[0]._id.toString(),
      sentTo: user._id.toString(),
      content: 'prova prova',
    };
    const res = await mutate(CREATE_MESSAGE, {
      variables: { message },
    });
    expect(res).toMatchSnapshot();
  });

  it('query messages succeds if logged in and admin, all are shown', async () => {
    const res = await query(MESSAGES);
    expect(res).toMatchSnapshot();
  });

  it('query messagesToUser succeds if logged in', async () => {
    const negotiation = await Negotiation.find({}).lean().exec();
    const res = await query(MESSAGES_USER, {
      variables: { user: negotiation[0].forUserAd.toString() },
    });
    expect(res).toMatchSnapshot();
  });

  it('query messagesForNegotiation succeds if logged in', async () => {
    const negotiation: LeanDocument<NegotiationDocument>[] =
      await Negotiation.find({}).lean().exec();
    const res = await query(MESSAGES_NEGOTIATION, {
      variables: { negotiation: negotiation[0]._id.toString() },
    });
    expect(res).toMatchSnapshot();
  });

  it('query single message succeds if logged', async () => {
    const message: LeanDocument<MessageDocument>[] = await Message.find({})
      .lean()
      .exec();
    const res = await query(MESSAGE, {
      variables: { id: message[0]._id.toString() },
    });
    expect(res).toMatchSnapshot();
  });

  it('query negotiations succeds if logged in and not admin, only user message shown', async () => {
    const data: any = await mutate(LOGIN_VALID_OTHER);
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    const { token } = data.data.login.response;
    setOptions({
      request: {
        headers: {
          authorization: token,
        },
      },
    });
    const res = await query(MESSAGES);
    expect(res).toMatchSnapshot();
  });
});

afterAll(async () => {
  await dropTestDb();
  await closeDbConnection();
});
