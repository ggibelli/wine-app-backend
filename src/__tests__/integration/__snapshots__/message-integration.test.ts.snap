// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration test messages create message mutation fails if not logged in 1`] = `
Object {
  "data": Object {
    "createMessage": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "locations": Array [
        Object {
          "column": 3,
          "line": 2,
        },
      ],
      "message": "You need to login to continue",
      "path": Array [
        "createMessage",
      ],
    },
  ],
}
`;

exports[`Integration test messages create message mutation fails if recipient same as sender 1`] = `
Object {
  "data": Object {
    "createMessage": Object {
      "errors": Array [
        Object {
          "name": "UserInputError",
          "text": "You cannot send messages to yourself",
        },
      ],
      "response": null,
    },
  },
}
`;

exports[`Integration test messages create message mutation fails if user does not own the negotiation 1`] = `
Object {
  "data": Object {
    "createMessage": Object {
      "errors": Array [
        Object {
          "name": "UserInputError",
          "text": "You can only send messages if you own the negotiation",
        },
      ],
      "response": null,
    },
  },
}
`;

exports[`Integration test messages create message mutation successfull 1`] = `
Object {
  "data": Object {
    "createMessage": Object {
      "errors": Array [],
      "response": Object {
        "content": "prova prova",
        "negotiation": Object {
          "createdBy": Object {
            "firstName": "Giovanni",
          },
          "forUserAd": Object {
            "firstName": "Mariuccio",
          },
        },
        "sentBy": Object {
          "firstName": "Giovanni",
        },
        "sentTo": Object {
          "firstName": "Mariuccio",
        },
      },
    },
  },
}
`;

exports[`Integration test messages query message fails if not logged in 1`] = `
Object {
  "data": Object {
    "messages": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "locations": Array [
        Object {
          "column": 3,
          "line": 2,
        },
      ],
      "message": "You need to login to continue",
      "path": Array [
        "messages",
      ],
    },
  ],
}
`;

exports[`Integration test messages query messageForNegotiation fails if not logged in 1`] = `
Object {
  "data": Object {
    "messagesForNegotiation": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "locations": Array [
        Object {
          "column": 3,
          "line": 2,
        },
      ],
      "message": "You need to login to continue",
      "path": Array [
        "messagesForNegotiation",
      ],
    },
  ],
}
`;

exports[`Integration test messages query messages succeds if logged in and admin, all are shown 1`] = `
Object {
  "data": Object {
    "messages": Array [
      Object {
        "content": "ciao",
        "sentBy": Object {
          "firstName": "Giovanni",
        },
        "sentTo": Object {
          "firstName": "Mariuccio",
        },
      },
      Object {
        "content": "ciao di nuovo",
        "sentBy": Object {
          "firstName": "Luigetto",
        },
        "sentTo": Object {
          "firstName": "Mariuccio",
        },
      },
      Object {
        "content": "prova prova",
        "sentBy": Object {
          "firstName": "Giovanni",
        },
        "sentTo": Object {
          "firstName": "Mariuccio",
        },
      },
    ],
  },
}
`;

exports[`Integration test messages query messagesForNegotiation succeds if logged in 1`] = `
Object {
  "data": Object {
    "messagesForNegotiation": Array [
      Object {
        "content": "ciao",
        "negotiation": Object {
          "createdBy": Object {
            "firstName": "Giovanni",
          },
          "forUserAd": Object {
            "firstName": "Mariuccio",
          },
        },
        "sentBy": Object {
          "firstName": "Giovanni",
        },
        "sentTo": Object {
          "firstName": "Mariuccio",
        },
      },
      Object {
        "content": "prova prova",
        "negotiation": Object {
          "createdBy": Object {
            "firstName": "Giovanni",
          },
          "forUserAd": Object {
            "firstName": "Mariuccio",
          },
        },
        "sentBy": Object {
          "firstName": "Giovanni",
        },
        "sentTo": Object {
          "firstName": "Mariuccio",
        },
      },
    ],
  },
}
`;

exports[`Integration test messages query messagesToUser fails if not logged in 1`] = `
Object {
  "data": Object {
    "messagesToUser": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "locations": Array [
        Object {
          "column": 3,
          "line": 2,
        },
      ],
      "message": "You need to login to continue",
      "path": Array [
        "messagesToUser",
      ],
    },
  ],
}
`;

exports[`Integration test messages query messagesToUser succeds if logged in 1`] = `
Object {
  "data": Object {
    "messagesToUser": Array [
      Object {
        "content": "ciao",
        "sentBy": Object {
          "firstName": "Giovanni",
        },
        "sentTo": Object {
          "firstName": "Mariuccio",
        },
      },
      Object {
        "content": "prova prova",
        "sentBy": Object {
          "firstName": "Giovanni",
        },
        "sentTo": Object {
          "firstName": "Mariuccio",
        },
      },
    ],
  },
}
`;

exports[`Integration test messages query negotiations succeds if logged in and not admin, only user message shown 1`] = `
Object {
  "data": Object {
    "messages": Array [
      Object {
        "content": "ciao",
        "sentBy": Object {
          "firstName": "Giovanni",
        },
        "sentTo": Object {
          "firstName": "Mariuccio",
        },
      },
      Object {
        "content": "ciao di nuovo",
        "sentBy": Object {
          "firstName": "Luigetto",
        },
        "sentTo": Object {
          "firstName": "Mariuccio",
        },
      },
      Object {
        "content": "prova prova",
        "sentBy": Object {
          "firstName": "Giovanni",
        },
        "sentTo": Object {
          "firstName": "Mariuccio",
        },
      },
    ],
  },
}
`;

exports[`Integration test messages query single message fails if not logged 1`] = `
Object {
  "data": Object {
    "message": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "locations": Array [
        Object {
          "column": 3,
          "line": 2,
        },
      ],
      "message": "You need to login to continue",
      "path": Array [
        "message",
      ],
    },
  ],
}
`;

exports[`Integration test messages query single message succeds if logged 1`] = `
Object {
  "data": Object {
    "message": Object {
      "content": "ciao",
      "sentBy": Object {
        "firstName": "Giovanni",
      },
      "sentTo": Object {
        "firstName": "Mariuccio",
      },
    },
  },
}
`;
